feat(node-details): add ability to view and edit Swarm mode nodes

Glowbal opened the pull request:
- Adds the ability to inspect a Swarm mode Node.
- Adds the ability to update a node's name, Availability and Labels.

Close #414

ncresswell commented:
Love it.. approved. Great work Glowbal....

lavvy commented:
Cool
On Jan 2, 2017 3:18 PM, "Neil Cresswell" <notifications@github.com> wrote:

> Love it.. approved. Great work Glowbal....
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/portainer/portainer/pull/417#issuecomment-270016078>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGsC4xm62Om-yQMJ8MLHysL4Eb2xiWU7ks5rOVuYgaJpZM4LV8ec>
> .
>

deviantony commented:
Really good work @Glowbal I love it too ! One thing that I dislike though: when loading a node view on a remote node (playwithdocker node for example) or with a poor internet connection, the view will display 'It looks like the node you wish to inspect does not exist.' for a short period of time before displaying the node view. IMO this should be replaced by a loading icon.

Glowbal commented:
I agree. I will make the change.

deviantony commented:
Awesome, ping me when you've made the changes as sometimes I don't receive a notification for PR updates.

Glowbal commented:
@deviantony I've made the changes. I've also fixed an error that occurred when you did a page refresh or opened the node details page as the first page. The `endpointMode` variable is not yet defined at that stage. To fix that, I've added recursive call with a timeout, until the variable is available.

I'm not sure if there is a better way to deal with that in AngularJS, but it does seem to work reliable.

deviantony commented on app/components/node/node.html:
```diff
@@ -0,0 +1,262 @@
+    </a>
+  </rd-header-title>
+  <rd-header-content>
+    <a ui-sref="swarm">Swarm Nodes</a> > <a ui-sref="node({id: node.Id})">{{ node.Id }}</a>
```
I think we should display the node hostname here instead of the node ID (easier to read in the navigation breadcrumb).

deviantony commented on app/components/node/nodeController.js:
```diff
@@ -0,0 +1,115 @@
+  };
+  $scope.removeLabel = function removeLabel(node, index) {
+    var removedElement = node.Labels.splice(index, 1);
+    if (removedElement != null) {
```
Please use `!==` for the comparison.

deviantony commented on app/components/node/nodeController.js:
```diff
@@ -0,0 +1,115 @@
+  function loadNodeAndTasks() {
+    // if endpointMode does not exist, this was a full page refresh or the first page visited.
+    // Wait a little until endpointmode has been determined
+    if (!$scope.endpointMode) {
```
I don't really like this solution. This problem is happening in other places in Portainer as well thanks for pointing it. I'll try to figure a better 'Angular' solution for this problem. In the meantime, please remove this check.

deviantony replied:
My work on this issue will be tracked in #522

Glowbal commented:
@deviantony I've made the requested changes.

deviantony commented on app/components/node/nodeController.js:
```diff
@@ -0,0 +1,107 @@
+angular.module('node', [])
+.controller('NodeController', ['$scope', '$state', '$stateParams', 'LabelHelper', 'Node', 'NodeHelper', 'Task', 'EndpointMode', 'Settings', 'Messages',
```
Remove any occurence of EndpointMode

deviantony replied:
EndpointMode service has been removed, please remove the dependency.

deviantony commented on app/components/node/nodeController.js:
```diff
@@ -0,0 +1,107 @@
+
+  function loadNodeAndTasks() {
+    $scope.loading = true;
+    if ($scope.endpointMode.provider === 'DOCKER_SWARM_MODE') {
```
You should now use `$scope.applicationState.endpoint.mode.provider` to check the associated Docker provider.

deviantony commented on app/shared/helpers.js:
```diff
@@ -1,4 +1,27 @@
 angular.module('portainer.helpers', [])
+
```
Please fix the conflict in this file
